server/server.go

```diff
@@ -281,6 +299,136 @@ func (h *reloadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	h.inner.ServeHTTP(w, r)
 }
 
+type JumpToDefResponse struct {
+	URL string `json:"url"`
+}
+
+func (s *server) ServeJumpToDef(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	docPositionParams, repo, err := s.parseDocPositionParams(r.URL.Query())
+	if err != nil {
+		writeError(ctx, w, 400, "bad_arguments", err.Error())
+		return
+	}
+
+	l := langserver.ForFile(repo, docPositionParams.TextDocument.URI)
+	langServer := s.langsrv[l.Address]
+	if langServer == nil {
+		// no language server
+		writeError(ctx, w, 404, "not_found", err.Error())
+		return
+	}
+
+	locations, err := langServer.Definition(ctx, docPositionParams)
+	if err != nil {
+		writeError(ctx, w, 500, "lsp_error", err.Error())
+		return
+	}
+
+	retUrl := ""
+
+	if len(locations) > 0 {
+		location := locations[0]
+		targetPath := strings.TrimPrefix(location.URI, "file://")
+		// Add 1 because URL is 1-indexed and language server is 0-indexed.
+		lineNum := location.Range.Start.Line + 1
+		if !strings.HasPrefix(targetPath, repo.Path) {
+			writeError(ctx, w, 400, "out_of_repo", "locations outside repo not supported")
+			return
+		}
+
+		relPath, err := filepath.Rel(repo.Path, targetPath)
+		if err != nil {
+			writeError(ctx, w, 500, "invalid_path", err.Error())
+			return
+		}
+
+		retUrl = fmt.Sprintf("/view/%s/%s#L%d", repo.Name, relPath, lineNum)
+	} else {
+		writeError(ctx, w, 400, "unresolved", "could not resolve an identifier")
+		return
+	}
+
+	replyJSON(ctx, w, 200, &JumpToDefResponse{
+		URL: retUrl,
+	})
+}
+
+func (s *server) ServeHover(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	docPositionParams, repo, err := s.parseDocPositionParams(r.URL.Query())
+	if err != nil {
+		writeError(ctx, w, 400, "bad_arguments", err.Error())
+		return
+	}
+
+	l := langserver.ForFile(repo, docPositionParams.TextDocument.URI)
+	langServer := s.langsrv[l.Address]
+	if langServer == nil {
+		// no language server
+		writeError(ctx, w, 404, "not_found", err.Error())
+		return
+	}
+
+	result, err := langServer.Hover(ctx, docPositionParams)
+	if err != nil {
+		writeError(ctx, w, 500, "lsp_error", err.Error())
+		return
+	}
+
+	replyJSON(ctx, w, 200, result)
+}
+
+func findRepo(repos []config.RepoConfig, repoName string) (*config.RepoConfig, error) {
+	for _, repo := range repos {
+		if repo.Name == repoName {
+			return &repo, nil
+		}
+	}
+	return nil, errors.New(fmt.Sprintf("repo with name %s not found", repoName))
+}
+
+func (s *server) parseDocPositionParams(params url.Values) (*langserver.TextDocumentPositionParams, *config.RepoConfig, error) {
+	row, column, repo, fpath :=
+		params.Get(rowParamName),
```

Comment:
We use `row` in the code, but LSP uses `line` -- is that a deliberate decision? I suspect it would be less confusing to standardize on `line`, or maybe `line_no` if we want to imply that it's a line number and not a string line of text.

Reply:
Done. Also switched to `character` instead of `col` to be more consistent with LSP.